### ISSUE TITLE
Deploy to docker

### ISSUE THREAD

**Issue by crosbymichael** [drone/drone#114]
So I'm creating an issue so that we can see if this is a good idea.  I have already starting working on this but want to make sure the config looks right.

#### So what I want out of a docker based CI system:
1. Get the git repo and do `docker build`, produce image
2. Run whatever test in a container from the image built
3. If tests pass, push image to a registry or the public index
4. Instruct a docker host running somewhere to pull down the image and run the container with a specified config.
5. Sit back and watch drone auto deploy services

This is what I'm thinking for the config.

```yaml
deploy:
  docker:
     tag: crosbymichael/skydock
     host: tcp://192.168.56.10:4243
     replace: skydock
     name: skydock
     volumes: ["/var/run/docker.sock:/docker.sock"]
     links: ["skydns:skydns"]
     cmd: ["-ttl", "30", "-environment", "dev", "-s" ,"/docker.sock", "-domain", "docker"]
```

A few problems that I see with drone is that it removes the images and containers after the build and that depoloyment scripts are just bash scripts that execute stuff and dont have information about the container and image in the build.  I would also like to consume the docker api via the client (in go) and not shell out on the cli.

**Comment by davidwindell** [comment]
I should think this issue can be closed now? #361 works a treat - even better when using the host's daemon to build the images which provides caching between builds.

**Comment by bradrydzewski** [comment]
@davidwindell I've kept this open because our support for building Docker images and publishing to the registry could be much better. The biggest issue is that you have to run your build in privileged mode. The good news is we've devised a solution and Docker and registry support should be seamless. Stay tuned.

**Comment by sunnysingh1985** [comment]
Hey @bradrydzewski , I have a couple of questions related to this issue, it would be great if could you please help me in understanding these:

1. Does current version of drone (I mean offficial version) have a way to build a docker container without me worrying about installing docker in docker?

2. I am understanding that you are saying that the current design needs to be improved but as of today I should atleast be able to get a container built using drone. Is this correct? And once the new design is implemented, I would just need to upgrade drone and things would not change much for me.

3. Is there a way currently to push the built docker images to dockerhub (given that you answer "yes" to my questions above :))

**Comment by davidwindell** [comment]
Hi @bradrydzewski, you can run drone on docker using our image https://registry.hub.docker.com/u/outeredge/edge-docker-drone/ if you like. All you need to do then is mount `/var/run/docker.sock` (no need for privileged mode). Your `.drone.yml` file will look something like this:

```yml
image: outeredge/edge-docker-drone-base
docker:
  net: host
script:
  - git-timestamp composer.json
  - git-timestamp composer.lock
publish:
    docker:
        docker_host: tcp://127.0.0.1:2377
        registry_login_url: $REGISTRY_URL
        registry_login: true
        keep_build: true
        username: $REGISTRY_USER
        password: $REGISTRY_PASS
        image_name: $IMAGE_NAME
        tags: [$DRONE_BRANCH, $(git rev-parse --short HEAD)]
        force_tags: true
```

**Comment by davidwindell** [comment]
One thing that's missing in this workflow is the ability to "test" the build image before pushing it to the registry.

**Comment by bradrydzewski** [comment]
Just merged the 0.4 branch into master (first major Drone release in a year, and more to follow). This release includes a pretty awesome container-based plugin model. We have an offiicial Docker plugin (see http://addons.drone.io/docker/) that will build and push your image to a public registry.

Enjoy!
